src/journal/journald-sync.c

```diff
@@ -0,0 +1,366 @@
+        assert(link);
+        assert(ret);
+
+        _cleanup_(sync_req_freep) SyncReq *req = new(SyncReq, 1);
+        if (!req)
+                return -ENOMEM;
+
+        *req = (SyncReq) {
+                .server = s,
+                .link = sd_varlink_ref(link),
+                .realtime_prioq_idx = PRIOQ_IDX_NULL,
+                .boottime_prioq_idx = PRIOQ_IDX_NULL,
+        };
+
+        /* We use five distinct mechanisms to determine when the synchronization request is complete:
+         *
+         * 1. For the syslog/native AF_UNIX/SOCK_DGRAM sockets we look at the datagram timestamps: once the
+         *    most recently seen datagram on the socket is newer than the timestamp when we initiated the
+         *    sync request we know that all previously enqueued messages have been processed by us.
+         *
+         * 2. For established stream AF_UNIX/SOCK_STREAM sockets we have no timestamps. For them we take the
+         *    SIOCINQ counter at the moment the synchronization request was enqueued. And once we processed
+         *    the indicated number of input bytes we know that anything further was enqueued later than the
+         *    original synchronization request we started from.
+         *
+         * 3. For pending new, un-accept()ed stream AF_UNIX/SOCK_STREAM sockets we have no timestamps either,
+         *    but we can query the number of pending connections via the sockdiag netlink protocol (I so wish
+         *    there was an easier, quicker way!). Once we accept()ed that many connections we know all
+         *    further connections are definitely more recent than the sync request.
+         *
+         * 4. For /dev/kmsg we look at the log message timestamps, similar to the AF_UNIX/SOCK_DGRAM case,
+         *    and they are in CLOCK_BOOTTIME clock.
+         *
+         * 5. Finally, as safety net we install an idle handler with a very low priority (lower than the
+         *    syslog/native/stream IO handlers). If this handler is called we know that there's no pending
+         *    IO, hence everything so far queued is definitely processed.
+         *
+         * Note the asymmetry: for AF_UNIX/SOCK_DGRAM + /dev/kmsg we go by timestamp, for established
+         * AF_UNIX/SOCK_STREAM we count bytes. That's because for SOCK_STREAM we have no timestamps, and for
+         * SOCK_DGRAM we have no API to query all pending bytes (as SIOCINQ on SOCK_DGRAM reports size of
+         * next datagram, not size of all pending datagrams). Ideally, we'd actually use neither of this, and
+         * the kernel would provide us CLOCK_MONOTONIC timestamps...
+         *
+         * Note that CLOCK_REALTIME is not necessarily monotonic (that's the whole point of it after all). If
+         * the clock jumps then we know the algorithm will eventually terminate, because of the idle handler
+         * that is our safety net. (Also, whenever we see poll() return an empty revents for some source we
```

Comment:
What happens if the clock jumps _and_ there is a steady, continuous stream of log messages, so that this never goes idle?

Reply:
so, if the time jumps forward, things are fine: the timestamps of the already queued messages will remain as they are, and remain older than the current time, hence we'll process them before returning.

if the time jumps backwards, things are not that great. we cannot relate the timestamps to the current time anymore, hence we'll quite likely not sync correctly: i.e. report the sync to be completed before all queued messages from before the sync are dispatched because we now think they are from the future. I'd say that is to be expected however, and we do finish in bounded time.

hence we should be good.

Reply:
> so, if the time jumps forward, things are fine: the timestamps of the already queued messages will remain as they are, and remain older than the current time, hence we'll process them before returning.

But if messages keep coming, then it will just keep processing them, rinse and repeat? How does it terminate in that case?

Reply:
if time jumps forward than all messages coming in later will have much newer timestamps, and hence we'll immediately finish the sync, and all is good.